### PR TITLE
[RFC] mux: check individual channels set and not entire row

### PR DESCRIPTION
**Pull request by juimonen**
Currently we are checking if matrix row contains any 1's (popcount). In
mux case this doesn't tell us if channels are mixed. For example if we
have 2 input matrix in which first row's position 2 and 4 have value 1,
it means that both streams channel 1 is copied to output channels 2 and
4, so not mixed. So change the check to be channel specific.

Signed-off-by: Jaska Uimonen <jaska.uimonen@intel.com>

### CODE REVIEW

**Comment by juimonen** [comment]
@bkokoszx @libinyang FYI

File: src/audio/mux/mux.c [hunk 1]
```diff
@@ -82,15 +83,17 @@ static int mux_set_values(struct comp_dev *dev, struct comp_data *cd,
...
-		for (j = 0 ; j < PLATFORM_MAX_CHANNELS; j++) {
-			channel_set = false;
-			for (i = 0 ; i < cfg->num_streams; i++) {
-				if (popcount(cfg->streams[i].mask[j])) {
-					if (!channel_set) {
-						channel_set = true;
-					} else {
-						comp_cl_err(&comp_mux, "mux_set_values(): mux component is not able to mix channels");
-						return -EINVAL;
+		for (i = 0 ; i < PLATFORM_MAX_CHANNELS; i++) {
```
Comment (lgirdwood) [inline src/audio/mux/mux.c:86]: Can we have a comment explaining this check.

**Review by lgirdwood — CHANGES_REQUESTED** [review]

**Comment by libinyang** [comment]
@juimonen when I apply your patch, It shows the below warning. I know we can use `--no-verify` to skip this warning. But it should be better to fix this warning, right?

```
WARNING: Too many leading tabs - consider code refactoring
#31: FILE: src/audio/mux/mux.c:91:
+                                               if (!channel_set) {

WARNING: Too many leading tabs - consider code refactoring
#33: FILE: src/audio/mux/mux.c:93:
+                                               } else {
```

**Review by bkokoszx — COMMENTED** [review]
@juimonen 
Thanks for this PR. I have one concern. Do we also check whether we perform mixing inside each individual input stream? I mean that we have e.g.:
```cfg->streams[0].mask[0] = 0x1```
```cfg->streams[0].mask[1] = 0x1```
It means that 0 and 1 channel from 0 stream goes to first channel at output, so there are also mixed. I think we do not check this case ATM.

**Comment by juimonen** [comment]
@bkokoszx yes you are right, let me try to make separate check functions for individual matrix and "inter-matrix", they are little bit nasty to combine and checkpatch goes also crazy for intendation level.

**Comment by juimonen** [comment]
@libinyang @bkokoszx new version, can you help to test and review?

File: src/audio/mux/mux.c [hunk 2]
```diff
@@ -44,12 +44,51 @@ DECLARE_SOF_RT_UUID("demux", demux_uuid, 0xc4b26868, 0x1430, 0x470e,
 
 DECLARE_TR_CTX(demux_tr, SOF_UUID(demux_uuid), LOG_LEVEL_INFO);
 
+static bool mux_mix_check(struct sof_mux_config *cfg)
```
Comment (lgirdwood) [inline src/audio/mux/mux.c:47]: Can we have a comment here describing at a high level what we are checking.

**Review by lgirdwood — CHANGES_REQUESTED** [review]

**Review by bkokoszx — APPROVED** [review]
Now, looks good to me. Thanks

**Comment by lgirdwood** [comment]
@juimonen can you check the CI

**Comment by juimonen** [comment]
SOFCI TEST

File: src/audio/mux/mux.c (hunk 1 above)
Comment (juimonen) [reply src/audio/mux/mux.c]: added something...

File: src/audio/mux/mux.c (hunk 2 above)
Comment (juimonen) [reply src/audio/mux/mux.c:54]: added something...

**Comment by juimonen** [comment]
updated: removed some rimage related bogus and added general comment to the check function

**Review by lgirdwood — APPROVED** [review]

**Pull request merged**